torch/csrc/jit/codegen/cuda/kernel_ir.h

```diff
@@ -22,12 +21,25 @@ namespace jit {
 namespace fuser {
 namespace kir {
 
+class IrBuilder;
+
+//! Token used to restrict the access to Kernel IR constructors
+//!
+//! Granular "friendship" token, used to implement the "passkey" idiom:
+//! https://www.spiria.com/en/blog/desktop-software/passkey-idiom-and-better-friendship-c
+//! https://arne-mertz.de/2016/10/passkey-idiom
+//!
+class Passkey {
```

Comment:
Just to make sure I understand this correctly, this is to enforce that IrBuilder is the only interface for creating KIR nodes, isn't it?

Reply:
Correct. We want to restrict the creation of IR nodes, so we don't have things like accidental `new kir::Int()` .

Private constructors + `friend` in the node themselves is an obvious alternative, although the passkey approach has a number of benefits:
1. It's granular - gives controlled access only to the constructors, not to everything private in the nodes
2. Preserves the logical "public" interface of the nodes (code readability)
3. It's more reliable than the raw C++ friend access (for ex. with passkey even the IR nodes can't create themselves w/o going through the IrBuilder interface)

Reply:
Yeah, interesting pattern. I'm glad I learned something new.